PhysicsTools/NanoAOD/plugins/DisTauTag.cc

```diff
@@ -0,0 +1,366 @@
+
+namespace {
+  template <typename T>
+  edm::Handle<T> getHandle(const edm::Event& event, const edm::EDGetTokenT<T>& token, bool get = true) {
+    edm::Handle<T> handle;
+    if (get)
+      event.getByToken(token, handle);
+    return handle;
+  }
+}  // namespace
+
+void test_vector(std::vector<float>& values) {
+  for (auto& value : values) {
+    if (std::isnan(value)) {
+      throw std::runtime_error("DisTauTag score output: NaN detected.");
+    } else if (std::isinf(value)) {
+      throw std::runtime_error("DisTauTag score output: Infinity detected.");
+    } else if (!std::isfinite(value)) {
+      throw std::runtime_error("DisTauTag score output: Non-standard value detected.");
+    }
+  }
+}
+
+class DisTauTag : public edm::global::EDProducer<> {
+public:
+  explicit DisTauTag(const edm::ParameterSet&);
+  ~DisTauTag() override { cleanupTensorFlow(); }
+
+  template <typename Scalar>
+  static Scalar getDeltaPhi(Scalar phi1, Scalar phi2);
+
+  static void fill_zero(tensorflow::Tensor&);
+
+private:
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  template <typename FeatureT>
+  const float Scale(const Int_t, const Float_t, const bool) const;
+  void saveInputs(const tensorflow::Tensor& tensor,
+                  const std::string& block_name,
+                  const edm::EventID& eventId,
+                  size_t jetIndex) const;
+
+  const std::string graphPath_;
+  const edm::EDGetTokenT<pat::JetCollection> jets_token;
+  const edm::EDGetTokenT<pat::PackedCandidateCollection> cands_token;
+  const bool save_inputs_;
+  const unsigned batch_size_;
+
+  // TensorFlow resources (mutable for thread safety)
+  mutable std::mutex session_mutex_;
+  mutable tensorflow::GraphDef* graphDef_;
+  mutable tensorflow::Session* session_;
+
+  void initializeTensorFlow() const;
+  void cleanupTensorFlow();
+};
+
+DisTauTag::DisTauTag(const edm::ParameterSet& config)
+    : graphPath_(config.getParameter<std::string>("graphPath")),
+      jets_token(consumes<pat::JetCollection>(config.getParameter<edm::InputTag>("jets"))),
+      cands_token(consumes<pat::PackedCandidateCollection>(config.getParameter<edm::InputTag>("pfCandidates"))),
+      save_inputs_(config.getParameter<bool>("save_inputs")),
+      batch_size_(config.getParameter<unsigned>("batchSize")),
+      graphDef_(nullptr),
+      session_(nullptr) {
+  produces<edm::ValueMap<float>>("score0");
+  produces<edm::ValueMap<float>>("score1");
+}
+
+void DisTauTag::initializeTensorFlow() const {
+  std::lock_guard<std::mutex> lock(session_mutex_);
+  if (!session_) {
+    if (!boost::filesystem::is_regular_file(graphPath_)) {
+      throw std::runtime_error("DisTauTag model not found: " + graphPath_);
+    }
+    graphDef_ = tensorflow::loadGraphDef(graphPath_);
+    session_ = tensorflow::createSession(graphDef_);
+  }
+}
+
+void DisTauTag::cleanupTensorFlow() {
+  std::lock_guard<std::mutex> lock(session_mutex_);
```

Comment:
The `lock` is not needed here. That would be more clear if the content of this function would be moved to the destructor (there is no way a destructor could be called concurrently).